### PR TITLE
samples: zbus: benchmark: complete refactoring and add msg subscriber to benchmark

### PR DESCRIPTION
**Pull request by rodrigopex**
> [!NOTE]  
> This PR rewrites the whole benchmark sample to improve code reuse and macro calls.

The benchmark sample did not contemplate message subscribers. It adds the msg subscribers and improves the organization of the project.

### CODE REVIEW

**Review by kartben — CHANGES_REQUESTED** [review]
some comments from my side -- fwiw it would help if the PR would be split in smaller chunks as there seems to be several changes here beyond just "adding message subscribers to benchmark", making the review more difficult than necessary. Thanks!

File: samples/subsys/zbus/benchmark/CMakeLists.txt [hunk 1]
```diff
@@ -4,5 +4,19 @@ cmake_minimum_required(VERSION 3.20.0)
...
+
+elseif(CONFIG_BM_SUBSCRIBERS)
+  list(APPEND app_sources src/sub.c)
+
+elseif(CONFIG_BM_MSG_SUBSCRIBERS)
+  list(APPEND app_sources src/msg_sub.c)
+
+endif()
+
+message(" > Benchmark files: ${app_sources}")
```
[the remark this answers is not in the document]
Comment (rodrigopex) [reply samples/subsys/zbus/benchmark/CMakeLists.txt:20]: I will remove it.

**Review by rodrigopex — COMMENTED** [review]
Add changes requested by @kartben.

**Comment by rodrigopex** [comment]
@kartben, is there anything I should change here to make that okay for merging?

**Comment by rodrigopex** [comment]
@kartben
> Hi @rodrigopex -- I will try to find time to review this but to clarify I was not asking for several **PRs** but rather to try and isolate the changes in several **commits** (but all in the same PR of course), so that reviewing the changes is easier. Here it looks like a fairly significant refactoring intertwined with what might be some less impactful changes, and it would have been great to be able to review the former while only quickly looking at the latter.

File: samples/subsys/zbus/benchmark/Kconfig [hunk 2]
```diff
@@ -9,8 +9,25 @@ config BM_ONE_TO
...
+	bool "Asynchronous using message subscribers"
+	select ZBUS_MSG_SUBSCRIBER
+
+endchoice
+
+config BM_FAIRPLAY
+	bool "Force a comparison with same actions"
+	help
+		Forces a message copy on the listeners and subscribers to behave equivalent to
+		message subscribers.
```
Comment (pdgendt) [inline samples/subsys/zbus/benchmark/Kconfig:31]: Nit
```suggestion
	help
	  Forces a message copy on the listeners and subscribers to behave equivalent to
	  message subscribers.
```

**Review by pdgendt — DISMISSED** [review]

File: samples/subsys/zbus/benchmark/Kconfig (hunk 2 above)
Comment (rodrigopex) [reply samples/subsys/zbus/benchmark/Kconfig]: Fixed.

**Review by pdgendt — DISMISSED** [review]

File: samples/subsys/zbus/benchmark/src/sub.c [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+				k_oops();                                                          \
+			}                                                                          \
+		}                                                                                  \
+	}                                                                                          \
+                                                                                                   \
+	K_THREAD_DEFINE(name##_id, CONSUMER_STACK_SIZE, name##_task, NULL, NULL, NULL, 3, 0, 0);
+
+#define CREATE_TASKS(s) S_TASK(s)
+FOR_EACH(CREATE_TASKS, (;), CONSUMERS_NAME_LIST)
+	;
```
Comment (fabiobaltieri) [inline samples/subsys/zbus/benchmark/src/sub.c:65]:
```suggestion
FOR_EACH(CREATE_TASKS, (;), CONSUMERS_NAME_LIST);
```
? For all of these I guess.

File: samples/subsys/zbus/benchmark/src/msg_sub.c [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+		struct bm_msg msg_received;                                                        \
+                                                                                                   \
+		while (1) {                                                                        \
+			if (zbus_sub_wait_msg(&name, &chan, &msg_received, K_FOREVER) == 0) {      \
+				atomic_add(&count, *((uint16_t *)msg_received.bytes));             \
+			} else {                                                                   \
+				k_oops();                                                          \
+			}                                                                          \
+		}                                                                                  \
+	}                                                                                          \
```
Comment (fabiobaltieri) [inline samples/subsys/zbus/benchmark/src/msg_sub.c:45]: Do you really have to generate these? Could you not  use the same code and `name` as argument?

File: samples/subsys/zbus/benchmark/src/sub.c (hunk 3 above)
Comment (fabiobaltieri) [reply samples/subsys/zbus/benchmark/src/sub.c:65]: yes please, let's not arbitrarily accept `clang-format` output, it's known to screw up macros, if you want to be able to rerun it just add a pair of clang format guards:

```
/* clang-format off */
..
/* clang-format on */
```

**Review by rodrigopex — COMMENTED** [review]
Changes requested have been attended, @fabiobaltieri.

**Review by pdgendt — DISMISSED** [review]

File: samples/subsys/zbus/benchmark/src/sub.c [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+#define CREATE_OBSERVATIONS(s) ZBUS_CHAN_ADD_OBS(bm_channel, s, 3)
+
+/* clang-format off */
+FOR_EACH(CREATE_OBSERVER, (;), CONSUMERS_NAME_LIST);
+
+FOR_EACH(CREATE_OBSERVATIONS, (;), CONSUMERS_NAME_LIST);
+/* clang-format on */
+
+int sub_thread(void *sub_ref, void *ptr2, void *ptr3)
```
Comment (fabiobaltieri) [inline samples/subsys/zbus/benchmark/src/sub.c:32]: `static`

File: samples/subsys/zbus/benchmark/src/sub.c [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+			}
+			zbus_chan_finish(chan);
+		} else {
+			k_oops();
+		}
+	}
+}
+
+#define S_TASK(name)                                                                               \
+	K_THREAD_DEFINE(name##_id, CONSUMER_STACK_SIZE, sub_thread, &name, NULL, NULL, 3, 0, 0);
```
[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply samples/subsys/zbus/benchmark/src/sub.c:64]: well, now that I look at the macro this one is debatable, let's skip it, or up to you I guess

File: samples/subsys/zbus/benchmark/src/sub.c (hunk 5 above)
Comment (rodrigopex) [reply samples/subsys/zbus/benchmark/src/sub.c]: Done.

**Review by rodrigopex — COMMENTED** [review]
Changes requested have been attended, @fabiobaltieri.

**Review by pdgendt — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Review by kartben — APPROVED** [review]

**Comment by rodrigopex** [comment]
Thank you for reviewing that, guys. Let's keep going!

**Pull request merged**